roachtest: sqlsmith: crdb_internal.reset_sql_stats(): sql stats resetter not set

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2812257&tab=buildLog) on [master@cbebc6e05491c6951216993ed5e12e22504624f2](https://github.com/cockroachdb/cockroach/commits/cbebc6e05491c6951216993ed5e12e22504624f2):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:226,sqlsmith.go:262,test_runner.go:768: error: pq: internal error: crdb_internal.reset_sql_stats(): sql stats resetter not set
		stmt:
		WITH
			with_642 (col_3897)
				AS (
					SELECT
						e'{"foo": [], "iqx;qV4": [0.475864628985448, null, [[false, "n<&\\"(p2M", []]], [null, []]]}':::JSONB
							AS col_3897
					FROM
						defaultdb.public.seed@[0] AS tab_1612
					WHERE
						true
				),
			with_648 (col_3909, col_3910)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								((), e']=c=;\x0e%':::STRING:::NAME),
								(
									(),
									CASE
									WHEN (false AND EXISTS (WITH with_643 (col_3898, col_3899, col_3900, col_3901, col_3902) AS (SELECT * FROM (VALUES ('M':::STRING, '0105000020E6100000070000000102000000070000001445AD92CCE657C0CD3D50F0495E53C0286036004C973240C0BA603ADD7853C0287FF36CDDF23940B6A5AF4AA1A14FC0A8DD7908CD814440EC1EE8DCCBFF51C0F2B2B8BFD9C16240E84B206501524040CAE83703FF245240A047CCC239423940986D807E21C25B40B46BF68F14E854400102000000030000005450A0E4F9CE60404F71B1D16EC450C03661EED88EE46240E00C48CB6AB729C030A888A0B26D5040F0C9DA828BDD4940010200000004000000C0C006825301364057AE6BC8BE1450C0F83A2DF8D7FD6540B68F9ACE005E3DC03CFB33234F554840805A3DF4A8C6F0BF741BD93DEB894AC0526F941ECE39514001020000000300000060C03BE4B55945405B7CD93B6D2A44C09C73F3F4892D5E40B0C0BDEF264312C02265CFC22C5762C0A8AA66A4F95749400102000000020000000E3AA152859B624043C9C4896B3942C0FA275EE96CE560407CA2CBBF169941C0010200000002000000E807A7C63B194540081F70F63E2650C0F8F906538CA33B405630D312B8BF3FC0010200000004000000E83F6E8065C36040DEC02F7D460A50C07EEE8963DD5F6340A0100D88471E0AC000AFF9B6CFE0554050566D92C7461E4000141A4ECF4943C0CC7EFB7646173740':::GEOGRAPHY, '{"DTh": null, "a": ["c", {}], "bar": {"foo": 0.021000040016623533}, "k#q}64z?JqwJ": {"$Ar": {"c": [{"a": []}]}, "ia|$]%MK_": {}}}':::JSONB, '5311d8dc-f3fc-4052-9297-ebc920ce93bd':::UUID, '2006-02-25 05:15:42.000523+00:00':::TIMESTAMPTZ), ('X':::STRING, '0101000020E61000008CA91246215B5940920E3ABF76985240':::GEOGRAPHY, '{"E`ssqK~": [["a", [], {}], [[null]], [null], {}], "bar": "bar", "foo": {}}':::JSONB, '1e5a0a57-fd63-4faf-8ad7-72d7028aa6b4':::UUID, '2008-03-07 19:08:22.000483+00:00':::TIMESTAMPTZ), (e'\x1e':::STRING, '01010000A0E61000003A6462E6E2DE6540E0CD53ACA04C05C0FE6C7110168CFBC1':::GEOGRAPHY, '[[], [false, "a"], {":/uo9:": {"c": ["foobar", [], {}]}, "foo": true}, [], [{"foo": {}}], {"6XH": true}]':::JSONB, '3d7e3645-d204-410d-8814-df01903bcbd2':::UUID, NULL), ('f':::STRING, '0107000020E6100000030000000105000020E610000004000000010200000009000000BC8055CA581A40C02713909A15E852C090F65E2648193FC0DBC9DF063BAC53C01AE3F33459426040399C2135E78A46C02880E67B33735940609AFEEB04A337C0F04D23EFED7E5D40A8CC9FB4195E3340ECDB9599857F5540AA43183229D55240C181D2A9712358C09C8B98DA2E414A405A7DF7675CC060C0B42322EBE47D414038765A21590661C0F445E6F3DA8E41400102000000020000001EDC7598ADAB634038C63A6E2C285640ACF5CA83EB674340C643D562904156400102000000020000000B1870ADEAE563C0208787FDEF8144409AA36163F97852C0880E5FA1DA0F2F40010200000004000000D824A57A47294940DCD0E1A3FE0F4FC06071D00168A356405C7E56602A0652C0ACDDF7B024295840FEF2917BC2F546C04C1E0A8A5E246440B83BECE44C494D400105000020E61000000800000001020000000300000018EA1F7ADC244EC020CB45B2ABB02B4038AD7A77E64D5F409095E384D8A1354080F49A10F9855640AC46222552424D40010200000004000000D4A69461C587634062A7288A77EA54C0984C886374A76540E0943852747548C0B8D061190DFE634068365D025A4049405C9B9AFD11045E40FC7400178DD7384001020000000300000078F3484FDEDA53400F2B1B5201164AC0B8BCFE5909DB5040EECF219A1E335140E07BB08991D832C0A4ED7DCB9B1C5540010200000002000000E8881DE9BC6E63409E71D89D7135424021F31DA1D7B165C0E0FD4EBDD6722D40010200000002000000D0C40730DDAC4F4088AA660F48A037C0931EBF60C0A863C04BF36C56822544C0010200000004000000847C96765FE458403EE304DE88AD42C07E9101D29C7265407C67C8A3CC7544C0920B21DB6A0B6440080E99CC4AA02340E0E062B07D194B402E3F877FB05F4140010200000003000000EE6E25D6AF616240F8CC45AA5B9152C0CCF3349F63C25A40009BA3FCB72FF83F9EFF60E05E536240B8CC693A20733E4001020000000300000008C9A47FFB0152C096F423DF0DE652401285F341714454C0F48B8E63FB8D3940D205834B3F7B54C00097A08B07AF2C400105000020E610000004000000010200000003000000E6D37E365FA155C03AAFC8FDD13151409A24DB7F7FD858C000CEE3F2863BEC3FDBCD82CAD2455BC040621CE0EB151BC0010200000004000000108BED5F04A63640109CC6D613584CC04C95F26BEA85484076E11561F85247C0DA68E00E5FC16440E0DF29C88CFB084050431FC7C2B748408C85E70450C43240010200000002000000FC3090AA16A849C0EC42A3471D6C40408126EB2612C364C09C51368799C84C40010200000002000000C509F6FA538963C0F173994BCBA855C000C0B08F6D2A0440C0DC87C8ED020640':::GEOGRAPHY, (e'[{"6i\\\\K": null, "t{w7CZlp-E": []}, [{"bar": {}, "oAmbPxUc": []}], {}, [], {"a": {}, "bar": []}]':::JSONB::JSONB - e'5\tdZ9':::STRING::STRING)::JSONB, '76641f60-c11b-48be-800b-c93115842548':::UUID, '1991-06-20 11:06:47.000276+00:00':::TIMESTAMPTZ)) AS tab_1613 (col_3898, col_3899, col_3900, col_3901, col_3902)), with_646 (col_3906) AS (SELECT * FROM (VALUES (B'10010000010101001110000110011001111001001100'), ((WITH with_644 (col_3903) AS (SELECT * FROM (VALUES ('05:34:53.155677':::TIME)) AS tab_1614 (col_3903)), with_645 (col_3904) AS (SELECT * FROM (VALUES (5895:::INT8), (21793:::INT8), (NULL), (12373:::INT8), (24859:::INT8), (0:::INT8)) AS tab_1615 (col_3904)) SELECT NULL AS col_3905 FROM defaultdb.public.seed@[0] AS tab_1616 WHERE tab_1616._bool ORDER BY tab_1616._uuid, tab_1616._int4 DESC, tab_1616._date DESC LIMIT 1:::INT8)), (B'00000000000000000000000000000000000000000000')) AS tab_1617 (col_3906)), with_647 (col_3907) AS (SELECT * FROM (VALUES (false), (true), (false)) AS tab_1618 (col_3907)) SELECT tab_1619._interval AS col_3908 FROM defaultdb.public.seed@[0] AS tab_1619 WHERE false))
									THEN e'c&\f\x0b9WX':::STRING:::NAME
									ELSE '{9v)!':::STRING:::NAME
									END
								),
								((), e'?_yeRns\x14A':::STRING:::NAME)
						)
							AS tab_1620 (col_3909, col_3910)
				)
		SELECT
			json_build_object(NULL)::JSONB AS col_3911
		FROM
			with_642 AS cte_ref_213
		GROUP BY
			cte_ref_213.col_3897
		HAVING
			crdb_internal.reset_sql_stats()::BOOL;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2812257&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/server/server_sql.go
+++ b/pkg/server/server_sql.go
@@ -599,6 +599,8 @@ func newSQLServer(ctx context.Context, cfg sqlServerArgs) (*SQLServer, error) {
 		execCfg,
 	)
 
+	distSQLServer.ServerConfig.SQLStatsResetter = pgServer.SQLServer
+
 	// Now that we have a pgwire.Server (which has a sql.Server), we can close a
 	// circular dependency between the rowexec.Server and sql.Server and set
 	// SessionBoundInternalExecutorFactory. The same applies for setting a

--- a/pkg/sql/distsql/server.go
+++ b/pkg/sql/distsql/server.go
@@ -299,6 +299,7 @@ func (ds *ServerImpl) setupFlow(
 			InternalExecutor:   ie,
 			Txn:                leafTxn,
 			SQLLivenessReader:  ds.ServerConfig.SQLLivenessReader,
+			SQLStatsResetter:   ds.ServerConfig.SQLStatsResetter,
 		}
 		evalCtx.SetStmtTimestamp(timeutil.Unix(0 /* sec */, req.EvalContext.StmtTimestampNanos))
 		evalCtx.SetTxnTimestamp(timeutil.Unix(0 /* sec */, req.EvalContext.TxnTimestampNanos))

--- a/pkg/sql/execinfra/server_config.go
+++ b/pkg/sql/execinfra/server_config.go
@@ -30,6 +30,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/hydratedtables"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlliveness"
 	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
 	"github.com/cockroachdb/cockroach/pkg/storage/cloud"
@@ -151,6 +152,10 @@ type ServerConfig struct {
 	// HydratedTables is a node-level cache of table descriptors which utilize
 	// user-defined types.
 	HydratedTables *hydratedtables.Cache
+
+	// SQLStatsResetter is an interface used to reset SQL stats without the need to
+	// introduce dependency on the sql package.
+	SQLStatsResetter tree.SQLStatsResetter
 }
 
 // RuntimeStats is an interface through which the rowexec layer can get

--- a/pkg/sql/logictest/testdata/logic_test/distsql_crdb_internal
+++ b/pkg/sql/logictest/testdata/logic_test/distsql_crdb_internal
@@ -1,0 +1,48 @@
+# LogicTest: 5node-default-configs
+
+statement ok
+CREATE TABLE data (a INT, b INT, c FLOAT, d DECIMAL, PRIMARY KEY (a, b, c, d))
+
+# Split into ten parts.
+statement ok
+ALTER TABLE data SPLIT AT SELECT i FROM generate_series(1, 9) AS g(i)
+
+# Relocate the ten parts to the five nodes.
+statement ok
+ALTER TABLE data EXPERIMENTAL_RELOCATE
+  SELECT ARRAY[i%5+1], i FROM generate_series(0, 9) AS g(i)
+
+# Generate all combinations of values 1 to 10.
+statement ok
+INSERT INTO data SELECT a, b, c::FLOAT, d::DECIMAL FROM
+   generate_series(1, 10) AS a(a),
+   generate_series(1, 10) AS b(b),
+   generate_series(1, 10) AS c(c),
+   generate_series(1, 10) AS d(d)
+
+# Verify data placement.
+query TTTI colnames,rowsort
+SELECT start_key, end_key, replicas, lease_holder FROM [SHOW RANGES FROM TABLE data]
+----
+start_key  end_key  replicas  lease_holder
+NULL       /1       {1}       1
+/1         /2       {2}       2
+/2         /3       {3}       3
+/3         /4       {4}       4
+/4         /5       {5}       5
+/5         /6       {1}       1
+/6         /7       {2}       2
+/7         /8       {3}       3
+/8         /9       {4}       4
+/9         NULL     {5}       5
+
+
+query II
+SELECT a, count(*) AS cnt FROM data GROUP BY a HAVING crdb_internal.reset_sql_stats() ORDER BY a LIMIT 5
+----
+1 1000
+2 1000
+3 1000
+4 1000
+5 1000
+

--- a/pkg/sql/planner.go
+++ b/pkg/sql/planner.go
@@ -414,6 +414,11 @@ func internalExtendedEvalCtx(
 ) extendedEvalContext {
 	evalContextTestingKnobs := execCfg.EvalContextTestingKnobs
 
+	var sqlStatsResetter tree.SQLStatsResetter
+	if execCfg.InternalExecutor != nil {
+		sqlStatsResetter = execCfg.InternalExecutor.s
+	}
+
 	return extendedEvalContext{
 		EvalContext: tree.EvalContext{
 			Txn:              txn,
@@ -428,6 +433,7 @@ func internalExtendedEvalCtx(
 			StmtTimestamp:    stmtTimestamp,
 			TxnTimestamp:     txnTimestamp,
 			InternalExecutor: execCfg.InternalExecutor,
+			SQLStatsResetter: sqlStatsResetter,
 		},
 		SessionMutator:    dataMutator,
 		VirtualSchemas:    execCfg.VirtualSchemas,

